e2e/test/MessageReceiveTests.cs

```diff
@@ -0,0 +1,456 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Azure.Devices.Client;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using System;
+using System.Diagnostics;
+using System.Diagnostics.Tracing;
+using System.Linq;
+using System.Net;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Azure.Devices.E2ETests
+{
+    [TestClass]
+    [TestCategory("IoTHub-E2E")]
+    public partial class MessageReceiveTests : IDisposable
+    {
+        private const string DevicePrefix = "E2E_Message_";
+        private const string DevicePrefixTimeout = "E2E_Message_Timeout_";
+        private static TestLogging _log = TestLogging.GetInstance();
+
+        private readonly ConsoleEventListener _listener;
+
+        public MessageReceiveTests()
+        {
+            _listener = new ConsoleEventListener("Microsoft-Azure-");
+        }
+
+        [TestMethod]
+        public async Task Message_DeviceReceiveSingleMessage_Amqp()
+        {
```

Comment:
Is there a better way to parameterize these tests? In Java, we just have a test for sending and receiving a single message, and the test suite passes in each protocol to use. Here you are still hardcoding all of the protocols to use, so you risk accidentally omitting a test case.

Reply:
It would also cut down on the space taken up by all the defined Message_DeviceReceiveSingleMessage_XXXXX methods here.

Reply:
We can easily do that when we switch to XUnit which supports data driven tests. 
I used to do this for VSTest but logging is very bad in that case so I had to roll-back.